server/src/main/java/io/druid/segment/realtime/firehose/EventReceiverFirehoseFactory.java

```diff
@@ -123,9 +139,30 @@ public EventReceiverFirehose(MapInputRowParser parser)
 
     @POST
     @Path("/push-events")
-    @Produces(MediaType.APPLICATION_JSON)
-    public Response addAll(Collection<Map<String, Object>> events)
+    @Consumes({MediaType.APPLICATION_JSON, SmileMediaTypes.APPLICATION_JACKSON_SMILE})
+    @Produces({MediaType.APPLICATION_JSON, SmileMediaTypes.APPLICATION_JACKSON_SMILE})
+    public Response addAll(
+        InputStream in,
+        @Context final HttpServletRequest req // used only to get request content-type
+    )
     {
+      final String reqContentType = req.getContentType();
+      final boolean isSmile = SmileMediaTypes.APPLICATION_JACKSON_SMILE.equals(reqContentType);
+      final String contentType = isSmile ? SmileMediaTypes.APPLICATION_JACKSON_SMILE : MediaType.APPLICATION_JSON;
+
+      ObjectMapper objectMapper = isSmile ? smileMapper : jsonMapper;
+
+      Collection<Map<String, Object>> events = null;
+      try {
+        events = objectMapper.readValue(
+            in, new TypeReference<Collection<Map<String, Object>>>()
+            {
+            }
+        );
+      }
+      catch (IOException e) {
+        Throwables.propagate(e);
```

Comment:
Can you wrap this? Maybe log locally and return just the message in a map of "error"->e.getMessage